### PR TITLE
Add ModStorage Lua API

### PR DESCRIPTION
**Pull request by nerzhul**
ModStorage Lua API is private per mod, if you want to make other mods access to it, expose functions 

* [x] Save data to world_dir/mod_storage (1 file per mod as a json)
* [x] Load data from world_dir/mod_storage
* [x] Save interval ? Save on modification ?
* [x] Use Metadata storage refactored by @rubenwardy

### CODE REVIEW

**Comment by rubenwardy** [comment]
how about `minetest.get_mod_storage()` at load time to get a mod's private store (getting mod name is reliable at load time when at root init.lua scope)

**Comment by nerzhul** [comment]
if we add mod_storage slot at mod load, you have arbitrary Lua and then you need to send this arbitrary variable to  read/write calls, except if you know a pure C++ mean to solve this problem. Lua mods should not know what is the mod storage

**Comment by numberZero** [comment]
Does that mean it uses `current_modname`? I’m definitely against that. A mod should always have access to its own storage, regardless of whether was its function called from some other mod or not.

**Comment by nerzhul** [comment]
yeah @numberZero it's not easy, we are trying to find a way to sandbox each mod

**Comment by numberZero** [comment]
>sandbox each mod

That’s impossible as long as you allow direct interaction of mods. Just as you can’t sandbox programs running in a shared memory.
In theory, it should be possible to do proper sandboxing. But it will be either slow or very complicated. Or probably both.

**Comment by ShadowNinja** [comment]
I'd like to do proper sandboxing on the client, but implementing that on the server would be a very breaking change.  At init time we have a decent guess of what the current mod is, but after that trying to guess is likely to cause more confusion than simplification (foo:node->on_place calls bar:dothing, which will now be accessing foo's mod store).

In short, the mod name has to be passed to the mod store functions.  This could be made less painful by getting a mod store object from something like `minetest.get_mod_store(mod_name)`.

**Comment by nerzhul** [comment]
yeah, then permit to use get_mod_storage() at mod launch, permitting it to get its own storage seems acceptable, this instanciate a coreside object reference, and after just manipulate it to store or load data

**Comment by nerzhul** [comment]
@ShadowNinja @rubenwardy there is now a ModStorage object which can only be fetch at mod load time. Only persistence is now missing

**Comment by celeron55** [comment]
Mod-specific sandboxing on the client requires each mod to have its own Lua environment and use a messaging system of some kind if they want to communicate with each other.

If that is the plan, then it has to be applied now, not later, because later it will be pretty much impossible just like on the server now.

**Comment by nerzhul** [comment]
@celeron55 yes it's possible to sandbox each mod(pack ?) in its own environment client side, but some things will be tricky, especially callbacks, and it's not the purpose of this PR

**Comment by sofar** [comment]
:+1: on the API (get mod storage handle ref, use it in further set/get calls).

File: doc/lua_api.txt [hunk 1]
```diff
@@ -2603,6 +2603,10 @@ These functions return the leftover itemstack.
 * `HTTPApiTable.fetch_async_get(handle)`: returns HTTPRequestResult
     * Return response data for given asynchronous HTTP request
 
+### Storage API:
+* `minetest.get_mod_storage()`:
+    * return reference to mod private `StorageRef`
```
Comment (rubenwardy) [inline doc/lua_api.txt]: > Must be called during load time

File: doc/lua_api.txt [hunk 2]
```diff
@@ -2803,6 +2807,14 @@ Can be gotten via `minetest.get_node_timer(pos)`.
 * `is_started()`: returns boolean state of timer
     * returns `true` if timer is started, otherwise `false`
 
+### `StorageRef`
+This is basically a reference to a C++ `ModStorage`
+
+#### Methods
+* `set(key, value)`: return true if key has been modified
```
Comment (rubenwardy) [inline doc/lua_api.txt]: returns

**Comment by nerzhul** [comment]
This is no longer a WIP. Mod storage is now properly working on server side and  can be merged/used

**Comment by nerzhul** [comment]
@sofar @sfan5 @rubenwardy please review a last time to make modders happy

**Comment by rubenwardy** [comment]
This should wait for my pr, imo

**Comment by nerzhul** [comment]
@rubenwardy which pr and why ?

**Comment by rubenwardy** [comment]
#5153 as this is basically meta

**Comment by ShadowNinja** [comment]
Er, this does nothing more than writing out a JSON file to the world directory, which mods can already do.  This doesn't seem to be useful when mods can just use a table and the serialization methods.

I thought this was for non-world-specific mod storage, which would actually be a new feature.

**Comment by numberZero** [comment]
@ShadowNinja, this adds new uniform API for mod storage. While currently it may use JSON files for that, it may be adopted to use a DB. It’s better to introduce working API now not waiting for the implementation to become perfect.
Also IMO mods should not have filesystem access, certainly not write access. This API removes need for that.
Just one notice, @nerzhul, it seems it supports strings only, why?

**Comment by nerzhul** [comment]
@numberZero it should support every type of objects like in player extended attributes, if you pass an object lua serialize it if i remember correctly

**Comment by nerzhul** [comment]
@rubenwardy it's rebased and now used the Metadata interfaces coreside & Lua side, as modified in your previous metadata storage. It works properly without changing many things, nice job !

**Comment by nerzhul** [comment]
@numberZero now since the refactor using Metadata it can store exactly what a Metadata can store, proper interface for all metadata, extensible for all core meta if needed

File: src/script/lua_api/l_storage.cpp [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+}
+
+Metadata* StorageRef::getmeta(bool auto_create)
+{
+	return m_object;
+}
+
+void StorageRef::clearMeta()
+{
+	// this function should be not used by mods
```
Comment (rubenwardy) [inline src/script/lua_api/l_storage.cpp]: this is needed to do `from_table`

File: src/metadata.cpp [hunk 4]
```diff
@@ -76,13 +76,27 @@ const std::string &Metadata::getString(const std::string &name,
...
-		m_stringvars[name] = var;
+		return true;
+	}
+
+	StringMap::iterator it = m_stringvars.find(name);
+	if (it != m_stringvars.end() && it->second == var) {
+		return false;
 	}
+
+	m_stringvars[name] = var;
```
Comment (nerzhul) [inline src/metadata.cpp]: as it can be m_stringvars.end() i'm not sure we can modify it, the spec is not clear about this an i prefer to be safe

**Comment by nerzhul** [comment]
@rubenwardy fixed, except one point as i'm not sure we can really do this, it's why i'm not using the iterator

File: src/mods.cpp [hunk 5]
```diff
@@ -356,3 +353,78 @@ Json::Value getModstoreUrl(std::string url)
...
+bool ModMetadata::save(const std::string &root_path)
+{
+	Json::Value json;
+	for (StringMap::const_iterator it = m_stringvars.begin();
+			it != m_stringvars.end(); ++it) {
+		json[it->first] = it->second;
+	}
+
+	if (!fs::PathExists(root_path)) {
+		fs::CreateAllDirs(root_path);
```
Comment (rubenwardy) [inline src/mods.cpp]: what if `CreateAllDirs` fails? Does `safeWriteToFile` just return false?

File: src/mods.cpp [hunk 6]
```diff
@@ -356,3 +353,78 @@ Json::Value getModstoreUrl(std::string url)
...
+			errorstream << "ModMetadata[" << m_mod_name << "]: Unable to save. '"
+				<< root_path << "' is not a directory." << std::endl;
+			return false;
+		}
+	}
+
+	bool w_ok = fs::safeWriteToFile(root_path + DIR_DELIM + m_mod_name,
+		Json::FastWriter().write(json));
+
+	if (!w_ok) {
```
Comment (rubenwardy) [inline src/mods.cpp]: would prefer the ! to be switched (ie, switch the blocks and invert the condition

File: src/mods.cpp [hunk 7]
```diff
@@ -356,3 +353,78 @@ Json::Value getModstoreUrl(std::string url)
...
+	Json::Value root;
+	if (!reader.parse(is, root)) {
+		errorstream << "ModMetadata[" << m_mod_name << "]: failed read data "
+			"(Json decoding failure)." << std::endl;
+		return false;
+	}
+
+	const Json::Value::Members attr_list = root.getMemberNames();
+	for (Json::Value::Members::const_iterator it = attr_list.begin();
+		it != attr_list.end(); ++it) {
```
Comment (rubenwardy) [inline src/mods.cpp]: double indent here

File: src/mods.cpp [hunk 8]
```diff
@@ -356,3 +353,78 @@ Json::Value getModstoreUrl(std::string url)
...
+		Json::Value attr_value = root[*it];
+		m_stringvars[*it] = attr_value.asString();
+	}
+
+	return true;
+}
+
+bool ModMetadata::setString(const std::string &name, const std::string &var)
+{
+	m_modified = Metadata::setString(name, var);
```
Comment (rubenwardy) [inline src/mods.cpp]: I'm wondering whether it may be worth making `m_modified` a property of Metadata, and removing the need for `virtual` instead

**Comment by sofar** [comment]
:+1:

**Comment by nerzhul** [comment]
I can port modification flag to parent class as a bool is tiny.

**Comment by nerzhul** [comment]
For doxygen there is no rule but it's nicer to have them with cpp permitting to read headers easily

**Comment by nerzhul** [comment]
Fixed some comments

File: src/mods.cpp (hunk 5 above)
Comment (nerzhul) [inline src/mods.cpp]: Fixed, it's exact we should handle this error, i missed it

**Comment by nerzhul** [comment]
@jdc843 the only possible case is a write failure when saving, or you modify the storage everytime, then it's flag as modified and wrote to disk

**Comment by jdc843** [comment]
"write failure when saving" - what is being saved and why is it being saved without any players?
I am the admin of Cash's World. I had just started a test server using the newly compiled MT software. I was about to log in when I noticed the messages.
server_map_save_interval = 5.3
map backend is SQLite3

Possibly related:
Something weird is happening that I don't know how to properly explain. When I log into a 0.4.15 server and do not have fly enabled, a 'noise' is repeated that has to do with me standing on something. The noise goes away when fly is enabled. This happens only when using mods that work fine using 0.4.14 server but goes away when I install the latest minetest_game.
Last thing: The world map for 0.4.15 seems to have changed. Once I run a 0.4.15 server using a map, I can't go back to using 0.4.14. The 0.4.14 server crashes complaining about the map format has changed.

I am learning a lot about how MT operates. But I still have a long way to go and the question I posted here might just be "stupid"

**Comment by Wayward1** [comment]
Regarding the third issue, the mapblock serialization version has recently been changed, as detailed here: https://forum.minetest.net/viewtopic.php?f=18&t=16657

**Pull request merged**